soroban-sdk/src/address.rs

```diff
@@ -1,13 +1,227 @@
+        Ok(v.to_raw())
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFrom<&Address> for ScVal {
+    type Error = ConversionError;
+    fn try_from(v: &Address) -> Result<Self, Self::Error> {
+        ScVal::try_from_val(&v.env, &v.obj.to_raw())
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFrom<Address> for ScVal {
+    type Error = ConversionError;
+    fn try_from(v: Address) -> Result<Self, Self::Error> {
+        (&v).try_into()
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFromVal<Env, ScVal> for Address {
+    type Error = ConversionError;
+    fn try_from_val(env: &Env, val: &ScVal) -> Result<Self, Self::Error> {
+        use soroban_env_host::TryIntoVal;
+        <_ as TryFromVal<_, Object>>::try_from_val(
+            env,
+            &val.try_into_val(env).map_err(|_| ConversionError)?,
+        )
+    }
+}
+
+impl Address {
+    /// Ensures that this Address has authorized invocation of the current
+    /// contract with the provided arguments.
+    ///
+    /// During the on-chain execution the Soroban host will perform the needed
+    /// authentication (verify the signatures) and ensure the replay prevention.
+    /// The contracts don't need to perform this tasks.
+    ///
+    /// The arguments don't have to match the arguments of the contract
+    /// invocation. However, it's considered the best practice to have a
+    /// well-defined, deterministic and ledger-state-independent mapping between
+    /// the contract invocation arguments and `require_auth` arguments. This
+    /// will allow the contract callers to easily build the required signature
+    /// payloads and prevent potential authorization failures.
+    ///
+    /// When called in the tests, the `require_auth` calls are just recorded and
+    /// no signatures are required. In order to make sure that the contract
+    /// has indeed called `require_auth` for this Address with expected arguments
+    /// use `env.verify_top_authorization`.
+    ///
+    /// ### Panics
+    ///
+    /// If the invocation is not authorized.
+    pub fn require_auth(&self, args: Vec<RawVal>) {
+        self.env.require_auth(&self, args);
+    }
+
+    #[inline(always)]
+    pub(crate) unsafe fn unchecked_new(env: Env, obj: Object) -> Self {
+        Self { env, obj }
+    }
+
+    #[inline(always)]
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+
+    pub fn as_raw(&self) -> &RawVal {
+        self.obj.as_raw()
+    }
+
+    pub fn to_raw(&self) -> RawVal {
+        self.obj.to_raw()
+    }
+
+    pub fn as_object(&self) -> &Object {
+        &self.obj
+    }
+
+    pub fn to_object(&self) -> Object {
+        self.obj
+    }
+}
+
+#[cfg(all(feature = "testutils", not(target_family = "wasm")))]
+use crate::env::xdr::{Hash, ScAddress, ScObject};
+#[cfg(all(feature = "testutils", not(target_family = "wasm")))]
+use crate::{testutils::random, BytesN};
+#[cfg(all(feature = "testutils", not(target_family = "wasm")))]
```

Comment:
The cfg's should be as any test and testutils I think, although it might not matter. That's what we used in other impl sections like this:

```suggestion
#[cfg(any(test, feature = "testutils"))]
use crate::env::xdr::{Hash, ScAddress, ScObject};
#[cfg(any(test, feature = "testutils"))]
use crate::{testutils::random, BytesN};
#[cfg(any(test, feature = "testutils"))]
```